xpotato/dataset/dataset.py

```diff
@@ -1,35 +1,109 @@
-import pickle
+from re import I
 from typing import List, Tuple, Dict
 
 import networkx as nx
 import pandas as pd
 
+from tqdm import tqdm
+from tuw_nlp.graph.utils import graph_to_pn
 from xpotato.dataset.sample import Sample
 from xpotato.graph_extractor.extract import GraphExtractor
+from xpotato.graph_extractor.graph import PotatoGraph
 
 
 class Dataset:
     def __init__(
-        self, examples: List[Tuple[str, str]], label_vocab: Dict[str, int], lang="en"
+        self,
+        examples: List[Tuple[str, str]] = None,
+        label_vocab: Dict[str, int] = {},
+        lang="en",
+        path=None,
+        binary=False,
     ) -> None:
         self.label_vocab = label_vocab
-        self._dataset = self.read_dataset(examples)
+        if path:
+            self._dataset = self.read_dataset(path=path, binary=binary)
+        else:
+            self._dataset = self.read_dataset(examples=examples)
         self.extractor = GraphExtractor(lang=lang, cache_fn=f"{lang}_nlp_cache")
         self.graphs = None
 
-    def read_dataset(self, examples: List[Tuple[str, str]]) -> List[Sample]:
-        return [Sample(example) for example in examples]
+    @staticmethod
+    def save_dataframe(df: pd.DataFrame, path: str) -> None:
+        graphs = [graph_to_pn(graph) for graph in df["graph"].tolist()]
+        df["graph"] = graphs
+        df.to_csv(path, index=False, sep="\t")
+
+    def prune_graphs(self, graphs: List[nx.DiGraph] = None) -> None:
+        graphs_str = []
+        for i, graph in enumerate(graphs):
+            graph.remove_nodes_from(list(nx.isolates(graph)))
+            # ADAM: THIS IS JUST FOR PICKLE TO PENMAN CONVERSION
+            graph = self._random_postprocess(graph)
+
+            g = [
+                c
+                for c in sorted(
+                    nx.weakly_connected_components(graph), key=len, reverse=True
+                )
+            ]
+            if len(g) > 1:
+                print(
+                    "WARNING: graph has multiple connected components, taking the largest"
```

Comment:
what about Kinga's solution for representing multiple connected components in penman? Is it too soon to use that?

Reply:
I think currently we don't have a good solution for the issue, but correct me if I'm wrong @GKingA